File: mobile/lib/core/websocket/websocket_client.dart
// ignore_for_file: dangling_library_doc_comments, unnecessary_library_name

/// WebSocket client for Synk mobile.
///
/// Provides a robust WebSocket connection to the Synk backend with:
/// - Authentication handshake
/// - Automatic reconnection with exponential backoff
/// - Connection state management
/// - Message parsing and dispatch
/// - Graceful disposal

import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/status.dart' as ws_status;

import 'package:synk_mobile/models/sync/sync_message.dart';

/// Connection state of the WebSocket client.
enum WebSocketConnectionState {
  disconnected,
  connecting,
  connected,
  reconnecting,
  error,
}

/// Configuration for WebSocket client behavior.
class WebSocketClientConfig {
  /// Initial reconnection delay in milliseconds.
  final int initialReconnectDelayMs;

  /// Maximum reconnection delay in milliseconds.
  final int maxReconnectDelayMs;

  /// Multiplier for exponential backoff.
  final double backoffMultiplier;

  /// Jitter factor (0.0 to 1.0) to prevent thundering herd.
  final double jitterFactor;

  /// Ping interval in milliseconds.
  final int pingIntervalMs;

  /// Connection timeout in milliseconds.
  final int connectionTimeoutMs;

  /// Maximum message size in bytes.
  final int maxMessageSizeBytes;

  const WebSocketClientConfig({
    this.initialReconnectDelayMs = 1000,
    this.maxReconnectDelayMs = 30000,
    this.backoffMultiplier = 2.0,
    this.jitterFactor = 0.3,
    this.pingIntervalMs = 30000,
    this.connectionTimeoutMs = 10000,
    this.maxMessageSizeBytes = 200000,
  });
}

/// Callback for connection state changes.
typedef ConnectionStateCallback = void Function(WebSocketConnectionState state);

/// Callback for received server messages.
typedef MessageCallback = void Function(ServerMessage message);

/// Callback for errors.
typedef ErrorCallback = void Function(Object error, StackTrace stackTrace);

/// WebSocket client for Synk backend.
///
/// Handles connection lifecycle, authentication, message parsing,
/// and automatic reconnection with exponential backoff.
class SynkWebSocketClient {
  final String _wsUrl;
  final String _deviceId;
  final WebSocketClientConfig _config;

  WebSocketChannel? _channel;
  StreamSubscription? _streamSubscription;
  Timer? _reconnectTimer;
  Timer? _pingTimer;
  Timer? _connectionTimeoutTimer;

  WebSocketConnectionState _state = WebSocketConnectionState.disconnected;
  int _reconnectAttempts = 0;
  bool _disposed = false;
  bool _authenticated = false;
  String? _lastError;

  final _stateController = StreamController<WebSocketConnectionState>.broadcast();
  final _messageController = StreamController<ServerMessage>.broadcast();
  final _errorController = StreamController<Object>.broadcast();

  /// Creates a new WebSocket client.
  ///
  /// [wsUrl] - WebSocket endpoint URL (e.g., ws://10.0.2.2:8000/api/v1/ws/{device_id})
  /// [deviceId] - Device ID for authentication
  /// [authToken] - Authentication token for handshake
  /// [config] - Optional configuration for reconnection and ping behavior
  SynkWebSocketClient({
    required this._wsUrl,
    required this._deviceId,
    required String authToken, // unused but kept for API compatibility
    WebSocketClientConfig? config,
  })  : _config = config ?? const WebSocketClientConfig();

  /// Current connection state.
  WebSocketConnectionState get state => _state;

  /// Stream of connection state changes.
  Stream<WebSocketConnectionState> get stateStream => _stateController.stream;

  /// Stream of received server messages.
  Stream<ServerMessage> get messageStream => _messageController.stream;

  /// Stream of errors.
  Stream<Object> get errorStream => _errorController.stream;

  /// Whether the client is currently connected and authenticated.
  bool get isConnected => _state == WebSocketConnectionState.connected && _authenticated;

  /// Last error message if any.
  String? get lastError => _lastError;

  /// Device ID for this connection.
  String get deviceId => _deviceId;

  /// Connects to the WebSocket server and performs authentication.
  Future<void> connect() async {
    if (_disposed) return;
    if (_state == WebSocketConnectionState.connecting ||
        _state == WebSocketConnectionState.connected) {
      return;
    }

    _setState(WebSocketConnectionState.connecting);
    _authenticated = false;

    try {
      final uri = Uri.parse(_wsUrl);
      _channel = WebSocketChannel.connect(uri);

      // Set up connection timeout
      _connectionTimeoutTimer = Timer(
        Duration(milliseconds: _config.connectionTimeoutMs),
        () {
          if (!_authenticated && _state == WebSocketConnectionState.connecting) {
            _handleConnectionError(
              Exception('Connection timeout during authentication'),
              StackTrace.current,
            );
          }
        },
      );

      // Listen for messages
      _streamSubscription = _channel!.stream.listen(
        _onMessage,
        onError: _onError,
        onDone: _onClose,
        cancelOnError: false,
      );
    } catch (e, st) {
      _handleConnectionError(e, st);
    }
  }

  /// Disconnects gracefully.
  Future<void> disconnect() async {
    _cancelReconnect();
    _cancelPingTimer();
    _cancelConnectionTimeout();

    if (_channel != null) {
      await _channel!.sink.close(ws_status.normalClosure);
      _channel = null;
    }

    if (_streamSubscription != null) {
      await _streamSubscription!.cancel();
      _streamSubscription = null;
    }

    _authenticated = false;
    _setState(WebSocketConnectionState.disconnected);
  }

  /// Sends a client message to the server.
  ///
  /// Returns true if the message was sent successfully.
  Future<bool> send(ClientMessage message) async {
    if (!isConnected || _channel == null) {
      return false;
    }

    try {
      final innerJson = message.when(
        clipboardUpdate: (m) => m.toJson(),
        ping: (m) => m.toJson(),
        deliveryReceipt: (m) => m.toJson(),
        historySyncRequest: (m) => m.toJson(),
      );
      final json = jsonEncode(innerJson);
      if (json.length > _config.maxMessageSizeBytes) {
        debugPrint('WebSocket message too large: ${json.length} bytes');
        return false;
      }
      _channel!.sink.add(json);
      return true;
    } catch (e, st) {
      _handleError(e, st);
      return false;
    }
  }

  /// Sends a clipboard update message.
  Future<bool> sendClipboardUpdate({
    required String messageId,
    required String text,
    ContentType contentType = ContentType.text,
  }) async {
    final message = ClipboardUpdateMessage(
      type: SyncMessageType.clipboardUpdate,
      messageId: messageId,
      deviceId: _deviceId,
      timestamp: DateTime.now().toUtc().toIso8601String().replaceAll('+00:00', 'Z'),
      payload: ClipboardPayload(
        contentType: contentType,
        text: text,
      ),
    );
    return send(ClientMessage.clipboardUpdate(message));
  }

  /// Sends a delivery receipt for a received clipboard update.
  Future<bool> sendDeliveryReceipt({
    required String messageId,
    required String receivedMessageId,
  }) async {
    final message = DeliveryReceiptMessage(
      type: SyncMessageType.deliveryReceipt,
      messageId: messageId,
      deviceId: _deviceId,
      timestamp: DateTime.now().toUtc().toIso8601String().replaceAll('+00:00', 'Z'),
      payload: {
        'received_message_id': receivedMessageId,
      },
    );
    return send(ClientMessage.deliveryReceipt(message));
  }

  /// Sends a history sync request.
  Future<bool> sendHistorySyncRequest({
    required String messageId,
    int limit = 50,
  }) async {
    final message = HistorySyncRequestMessage(
      type: SyncMessageType.historySyncRequest,
      messageId: messageId,
      deviceId: _deviceId,
      timestamp: DateTime.now().toUtc().toIso8601String().replaceAll('+00:00', 'Z'),
      payload: {
        'limit': limit,
      },
    );
    return send(ClientMessage.historySyncRequest(message));
  }

  /// Sends a ping message.
  Future<bool> sendPing({String? messageId}) async {
    final message = PingMessage(
      type: SyncMessageType.ping,
      messageId: messageId ?? _generateMessageId(),
      deviceId: _deviceId,
      timestamp: DateTime.now().toUtc().toIso8601String().replaceAll('+00:00', 'Z'),
    );
    return send(ClientMessage.ping(message));
  }

  /// Disposes all resources.
  void dispose() {
    _disposed = true;
    disconnect();
    _stateController.close();
    _messageController.close();
    _errorController.close();
  }

  void _setState(WebSocketConnectionState newState) {
    if (_state != newState) {
      _state = newState;
      _stateController.add(newState);
    }
  }

  void _onMessage(dynamic data) {
    if (_disposed) return;

    if (data is! String) {
      debugPrint('Received non-string WebSocket message: ${data.runtimeType}');
      return;
    }

    if (data.length > _config.maxMessageSizeBytes) {
      debugPrint('Received message too large: ${data.length} bytes');
      return;
    }

    try {
      final json = jsonDecode(data) as Map<String, dynamic>;
      final serverMessage = ServerMessage.fromJson(json);

      // Handle authentication success
      if (serverMessage is ServerMessageAuthSuccess) {
        _authenticated = true;
        _cancelConnectionTimeout();
        _startPingTimer();
        _setState(WebSocketConnectionState.connected);
        _reconnectAttempts = 0;
        debugPrint('WebSocket authenticated successfully');
      }

      _messageController.add(serverMessage);
    } catch (e, st) {
      // Log but don't crash on malformed messages
      if (kDebugMode) {
        debugPrint('Failed to parse WebSocket message: $e');
        debugPrint('Raw message: $data');
      }
      _handleError(
        FormatException('Failed to parse WebSocket message: $e'),
        st,
      );
    }
  }

  void _onError(Object error) {
    if (_disposed) return;
    _handleError(error, StackTrace.current);
  }

  void _onClose() {
    if (_disposed) return;

    _cancelConnectionTimeout();
    _cancelPingTimer();
    _authenticated = false;

    if (_state != WebSocketConnectionState.disconnected) {
      _setState(WebSocketConnectionState.disconnected);
      _scheduleReconnect();
    }
  }

  void _handleConnectionError(Object error, StackTrace stackTrace) {
    if (_disposed) return;

    _lastError = error.toString();
    _cancelConnectionTimeout();
    _setState(WebSocketConnectionState.error);
    _errorController.add(error);

    if (_channel != null) {
      _channel!.sink.close(ws_status.abnormalClosure);
      _channel = null;
    }

    if (_streamSubscription != null) {
      _streamSubscription!.cancel();
      _streamSubscription = null;
    }

    _scheduleReconnect();
  }

  void _handleError(Object error, StackTrace stackTrace) {
    if (_disposed) return;
    _lastError = error.toString();
    _errorController.add(error);
  }

  void _scheduleReconnect() {
    if (_disposed) return;
    if (_state == WebSocketConnectionState.reconnecting) return;

    _setState(WebSocketConnectionState.reconnecting);

    // Calculate delay with exponential backoff and jitter
    final baseDelay = min(
      _config.initialReconnectDelayMs * pow(_config.backoffMultiplier, _reconnectAttempts),
      _config.maxReconnectDelayMs,
    ).toInt();

    final jitter = (baseDelay * _config.jitterFactor * Random().nextDouble()).toInt();
    final delay = baseDelay + jitter;

    _reconnectAttempts++;
    debugPrint('Scheduling WebSocket reconnect in ${delay}ms (attempt $_reconnectAttempts)');

    _reconnectTimer = Timer(Duration(milliseconds: delay), () {
      if (!_disposed) {
        connect();
      }
    });
  }

  void _startPingTimer() {
    _cancelPingTimer();
    _pingTimer = Timer.periodic(
      Duration(milliseconds: _config.pingIntervalMs),
      (_) {
        if (isConnected) {
          sendPing();
        }
      },
    );
  }

  void _cancelPingTimer() {
    _pingTimer?.cancel();
    _pingTimer = null;
  }

  void _cancelConnectionTimeout() {
    _connectionTimeoutTimer?.cancel();
    _connectionTimeoutTimer = null;
  }

  void _cancelReconnect() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
  }

  String _generateMessageId() {
    return DateTime.now().microsecondsSinceEpoch.toRadixString(36) +
        Random().nextInt(1000000).toRadixString(36);
  }
}

File: desktop/src/main/services/websocket-client.ts
/**
 * WebSocket client for Synk desktop.
 *
 * Provides a robust WebSocket connection to the Synk backend with:
 * - Authentication handshake
 * - Automatic reconnection with exponential backoff
 * - Connection state management
 * - Message parsing and dispatch
 * - Graceful disposal
 */

import { EventEmitter } from 'events';
import WebSocket from 'ws';
import {
  SyncMessageType,
  ContentType,
  ClipboardPayload,
  ClientMessage,
  ServerMessage,
  ClipboardUpdateMessage,
  PingMessage,
  DeliveryReceiptMessage,
  HistorySyncRequestMessage,
  ClipboardUpdateRelayMessage,
  ErrorMessage,
  AckMessage,
  HistorySyncResponseMessage,
  DeviceOnlineMessage,
  DeviceOfflineMessage,
  AuthSuccessMessage,
  ClipboardUpdateReceivedMessage,
  PongMessage,
  parseServerMessage,
  generateMessageId,
  generateTimestamp,
} from '../models/sync';
import { SyncMessageBase } from '../models/sync-base';

/** Connection state of the WebSocket client. */
export enum WebSocketConnectionState {
  Disconnected = 'disconnected',
  Connecting = 'connecting',
  Connected = 'connected',
  Reconnecting = 'reconnecting',
  Error = 'error',
}

/** Configuration for WebSocket client behavior. */
export interface WebSocketClientConfig {
  /** Initial reconnection delay in milliseconds. */
  initialReconnectDelayMs: number;
  /** Maximum reconnection delay in milliseconds. */
  maxReconnectDelayMs: number;
  /** Multiplier for exponential backoff. */
  backoffMultiplier: number;
  /** Jitter factor (0.0 to 1.0) to prevent thundering herd. */
  jitterFactor: number;
  /** Ping interval in milliseconds. */
  pingIntervalMs: number;
  /** Connection timeout in milliseconds. */
  connectionTimeoutMs: number;
  /** Maximum message size in bytes. */
  maxMessageSizeBytes: number;
}

/** Default configuration. */
export const DEFAULT_CONFIG: WebSocketClientConfig = {
  initialReconnectDelayMs: 1000,
  maxReconnectDelayMs: 30000,
  backoffMultiplier: 2.0,
  jitterFactor: 0.3,
  pingIntervalMs: 30000,
  connectionTimeoutMs: 10000,
  maxMessageSizeBytes: 200000,
};

/** Events emitted by the WebSocket client. */
export interface SynkWebSocketClientEvents {
  stateChange: [state: WebSocketConnectionState];
  message: [message: ServerMessage];
  error: [error: Error];
}

/**
 * WebSocket client for Synk backend.
 *
 * Handles connection lifecycle, authentication, message parsing,
 * and automatic reconnection with exponential backoff.
 */
export class SynkWebSocketClient extends EventEmitter<SynkWebSocketClientEvents> {
  private readonly wsUrl: string;
  private readonly deviceId: string;
  private readonly config: WebSocketClientConfig;

  private ws: WebSocket | null = null;
  private reconnectTimer: NodeJS.Timeout | null = null;
  private pingTimer: NodeJS.Timeout | null = null;
  private connectionTimeoutTimer: NodeJS.Timeout | null = null;

  private state: WebSocketConnectionState = WebSocketConnectionState.Disconnected;
  private reconnectAttempts = 0;
  private disposed = false;
  private authenticated = false;
  private lastError: string | null = null;

  /**
   * Creates a new WebSocket client.
   *
   * @param wsUrl - WebSocket endpoint URL (e.g., ws://10.0.2.2:8000/api/v1/ws/{device_id})
   * @param deviceId - Device ID for authentication
   * @param authToken - Authentication token for handshake (sent in first message body)
   * @param config - Optional configuration for reconnection and ping behavior
   */
  constructor(
    wsUrl: string,
    deviceId: string,
    private readonly authToken: string,
    config?: Partial<WebSocketClientConfig>
  ) {
    super();
    this.wsUrl = wsUrl;
    this.deviceId = deviceId;
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  /** Current connection state. */
  get connectionState(): WebSocketConnectionState {
    return this.state;
  }

  /** Whether the client is currently connected and authenticated. */
  get isConnected(): boolean {
    return this.state === WebSocketConnectionState.Connected && this.authenticated;
  }

  /** Last error message if any. */
  get lastErrorMessage(): string | null {
    return this.lastError;
  }

  /** Device ID for this connection. */
  get deviceIdGetter(): string {
    return this.deviceId;
  }

  /**
   * Connects to the WebSocket server and performs authentication.
   */
  async connect(): Promise<void> {
    if (this.disposed) return;
    if (
      this.state === WebSocketConnectionState.Connecting ||
      this.state === WebSocketConnectionState.Connected
    ) {
      return;
    }

    this.setState(WebSocketConnectionState.Connecting);
    this.authenticated = false;

    try {
      this.ws = new WebSocket(this.wsUrl);

      // Set up connection timeout
      this.connectionTimeoutTimer = setTimeout(() => {
        if (!this.authenticated && this.state === WebSocketConnectionState.Connecting) {
          this.handleConnectionError(
            new Error('Connection timeout during authentication')
          );
        }
      }, this.config.connectionTimeoutMs);

      // Listen for events
      this.ws.on('open', () => this.onOpen());
      this.ws.on('message', (data: Buffer) => this.onMessage(data));
      this.ws.on('error', (error: Error) => this.onError(error));
      this.ws.on('close', (code: number, reason: Buffer) =>
        this.onClose(code, reason.toString())
      );
    } catch (error) {
      this.handleConnectionError(error as Error);
    }
  }

  /**
   * Disconnects gracefully.
   */
  async disconnect(): Promise<void> {
    this.cancelReconnect();
    this.cancelPingTimer();
    this.cancelConnectionTimeout();

    // Set state to Disconnected BEFORE closing WebSocket to prevent onClose from scheduling reconnect
    this.authenticated = false;
    this.setState(WebSocketConnectionState.Disconnected);

    if (this.ws) {
      this.ws.close(1000, 'Normal closure');
      this.ws = null;
    }
  }

  /**
   * Sends a client message to the server.
   *
   * @returns true if the message was sent successfully
   */
  async send(message: ClientMessage): Promise<boolean> {
    if (!this.isConnected || !this.ws) {
      return false;
    }

    try {
      const json = JSON.stringify(message);
      if (json.length > this.config.maxMessageSizeBytes) {
        return false;
      }
      this.ws.send(json);
      return true;
    } catch (error) {
      this.handleError(error as Error);
      return false;
    }
  }

  /**
   * Sends a clipboard update message.
   */
  async sendClipboardUpdate(
    messageId: string,
    text: string,
    contentType: ContentType = ContentType.TEXT
  ): Promise<boolean> {
    const message: ClipboardUpdateMessage = {
      type: SyncMessageType.CLIPBOARD_UPDATE,
      version: 1,
      message_id: messageId,
      device_id: this.deviceId,
      timestamp: generateTimestamp(),
      payload: {
        content_type: contentType,
        text,
      },
    };
    return this.send(message);
  }

  /**
   * Sends a delivery receipt for a received clipboard update.
   */
  async sendDeliveryReceipt(
    messageId: string,
    receivedMessageId: string
  ): Promise<boolean> {
    const message: DeliveryReceiptMessage = {
      type: SyncMessageType.DELIVERY_RECEIPT,
      version: 1,
      message_id: messageId,
      device_id: this.deviceId,
      timestamp: generateTimestamp(),
      payload: {
        received_message_id: receivedMessageId,
      },
    };
    return this.send(message);
  }

  /**
   * Sends a history sync request.
   */
  async sendHistorySyncRequest(
    messageId: string,
    limit: number = 50
  ): Promise<boolean> {
    const message: HistorySyncRequestMessage = {
      type: SyncMessageType.HISTORY_SYNC_REQUEST,
      version: 1,
      message_id: messageId,
      device_id: this.deviceId,
      timestamp: generateTimestamp(),
      payload: {
        limit,
      },
    };
    return this.send(message);
  }

  /**
   * Sends a ping message.
   */
  async sendPing(messageId?: string): Promise<boolean> {
    const message: PingMessage = {
      type: SyncMessageType.PING,
      version: 1,
      message_id: messageId ?? generateMessageId(),
      device_id: this.deviceId,
      timestamp: generateTimestamp(),
      payload: undefined,
    };
    return this.send(message);
  }

  /**
   * Disposes all resources.
   */
  dispose(): void {
    this.disposed = true;
    this.disconnect();
    this.removeAllListeners();
  }

  private setState(newState: WebSocketConnectionState): void {
    if (this.state !== newState) {
      this.state = newState;
      this.emit('stateChange', newState);
    }
  }

  private onOpen(): void {
    if (this.disposed) return;

    // Send authentication message as first message
    const authMessage = {
      type: 'auth',
      token: this.authToken,
    };
    this.ws?.send(JSON.stringify(authMessage));
  }

  private onMessage(data: Buffer): void {
    if (this.disposed) return;

    const messageStr = data.toString();
    if (messageStr.length > this.config.maxMessageSizeBytes) {
      return;
    }

    try {
      const serverMessage = parseServerMessage(messageStr);

      // Handle authentication success
      if (isAuthSuccessMessage(serverMessage)) {
        this.authenticated = true;
        this.cancelConnectionTimeout();
        this.startPingTimer();
        this.setState(WebSocketConnectionState.Connected);
        this.reconnectAttempts = 0;
      }

      this.emit('message', serverMessage);
    } catch (error) {
      this.handleError(
        new Error(`Failed to parse WebSocket message: ${error}`)
      );
    }
  }

  private onError(error: Error): void {
    if (this.disposed) return;
    // WebSocket errors are connection errors - handle them as such
    this.handleConnectionError(error);
  }

  private onClose(code: number, reason: string): void {
    if (this.disposed) return;

    this.cancelConnectionTimeout();
    this.cancelPingTimer();
    this.authenticated = false;

    // Don't schedule reconnect if we're already in Error state (connection error already scheduled one)
    // Or if we're explicitly disconnecting (state = Disconnected)
    if (this.state !== WebSocketConnectionState.Disconnected &&
        this.state !== WebSocketConnectionState.Error) {
      this.setState(WebSocketConnectionState.Disconnected);
      this.scheduleReconnect();
    } else if (this.state === WebSocketConnectionState.Error) {
      // Already in error state, just update to Disconnected but don't schedule another reconnect
      this.setState(WebSocketConnectionState.Disconnected);
    }
  }

  private handleConnectionError(error: Error): void {
    if (this.disposed) return;

    this.lastError = error.message;
    this.cancelConnectionTimeout();
    this.setState(WebSocketConnectionState.Error);
    this.emit('error', error);

    if (this.ws) {
      this.ws.close(1006, 'Abnormal closure');
      this.ws = null;
    }
    // Reconnect will be scheduled in onClose when the WebSocket actually closes
  }

  private handleError(error: Error): void {
    if (this.disposed) return;
    this.lastError = error.message;
    this.emit('error', error);
  }

  private scheduleReconnect(): void {
    if (this.disposed) return;
    if (this.state === WebSocketConnectionState.Reconnecting) return;

    this.setState(WebSocketConnectionState.Reconnecting);

    // Calculate delay with exponential backoff and jitter
    const baseDelay = Math.min(
      this.config.initialReconnectDelayMs *
        Math.pow(this.config.backoffMultiplier, this.reconnectAttempts),
      this.config.maxReconnectDelayMs
    );

    const jitter = Math.floor(
      baseDelay * this.config.jitterFactor * Math.random()
    );
    const delay = baseDelay + jitter;

    this.reconnectAttempts++;
    console.log(
      `Scheduling WebSocket reconnect in ${delay}ms (attempt ${this.reconnectAttempts})`
    );

    this.reconnectTimer = setTimeout(() => {
      if (!this.disposed) {
        this.connect();
      }
    }, delay);
  }

  private startPingTimer(): void {
    this.cancelPingTimer();
    this.pingTimer = setInterval(() => {
      if (this.disposed) return;
      if (this.isConnected) {
        this.sendPing();
      }
    }, this.config.pingIntervalMs);
  }

  private cancelPingTimer(): void {
    if (this.pingTimer) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
  }

  private cancelConnectionTimeout(): void {
    if (this.connectionTimeoutTimer) {
      clearTimeout(this.connectionTimeoutTimer);
      this.connectionTimeoutTimer = null;
    }
  }

  private cancelReconnect(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
  }
}

/** Type guard functions for message discrimination. */
export function isAuthSuccessMessage(msg: ServerMessage): msg is AuthSuccessMessage {
  return msg.type === SyncMessageType.AUTH_SUCCESS;
}

export function isClipboardUpdateRelay(msg: ServerMessage): msg is ClipboardUpdateRelayMessage {
  return msg.type === SyncMessageType.CLIPBOARD_UPDATE_RELAY;
}

export function isErrorMessage(msg: ServerMessage): msg is ErrorMessage {
  return msg.type === SyncMessageType.ERROR;
}

export function isAckMessage(msg: ServerMessage): msg is AckMessage {
  return msg.type === SyncMessageType.ACK;
}

export function isHistorySyncResponse(msg: ServerMessage): msg is HistorySyncResponseMessage {
  return msg.type === SyncMessageType.HISTORY_SYNC_RESPONSE;
}

export function isDeviceOnlineMessage(msg: ServerMessage): msg is DeviceOnlineMessage {
  return msg.type === SyncMessageType.DEVICE_ONLINE;
}

export function isDeviceOfflineMessage(msg: ServerMessage): msg is DeviceOfflineMessage {
  return msg.type === SyncMessageType.DEVICE_OFFLINE;
}

export function isPongMessage(msg: ServerMessage): msg is PongMessage {
  return msg.type === SyncMessageType.PONG;
}

export function isClipboardUpdateReceived(msg: ServerMessage): msg is ClipboardUpdateReceivedMessage {
  return msg.type === SyncMessageType.CLIPBOARD_UPDATE_RECEIVED;
}